docs/specs/src/hwp_4_formal.md

```diff
@@ -0,0 +1,403 @@
+# Holochain Formal Design Elements
+
+Now we turn to a more formal and detailed presentation of the Holochain
+system, including assumptions, architecture, integrity guarantees, and
+formal state model.
+
+**Purpose of this Section:** To provide an understanding of the functional requirements of Holochain and specify a technical implmentation of the cryptographic state transitions and application processes that enforce Holochain's integrity guarantees. 
+
+## Definition of Foundational Principles
+- **Cryptography:** Holochain's integrity guarantees are largely enabled by cryptography. It is used in three main ways.
+     - *Hashes:* Data is uniquely identified by its hash which is the key used to retrieve from Content Addressable Stores. 
+     - *Signing:* Origination of data (for all storage and network communications) is verified by signing of a hash with a private key.
+     - *Encryption:* Data is encrypted at rest and on the wire throughout the system.
+- **Agency:** Holochain is agent-centric. Each and every state change is a result of: 1) record of an agent's action 2) signed by the authoring agent, 3) linearly sequenced and timestamped 4) to their local source chain. Each agent is the sole authority for managing its local state (by virtue of controlling their private key required for signing new entries to a hashchain).
+- **Accountability:** Holochain is also socio-centric. Each Holochain application defines its set of mutually enforced data integrity rules. Every local state change gets validated by other agents to ensure that it adhered to the rules of that application. Peers also enforce limits on publishing rates, protect against network integrity threats, and can ban rule-breakers by a process we call *warranting*.
+- **Data:** Unlike most other decentralized approaches, in Holochain, data does not have first-order, independent, ontological existence. Every record in the shared DHT network space MUST CARRY its provenance from a local source chain as described below. 
+- **Provenance:** Each record created in a Holochain application starts as an entry/action pair on someone's local source chain. As such, even when published to the shared DHT, entries must carry the associated action signed by the agent who created it. This means every piece of data carries meta information about where that data came from (who created it in what sequence on the their chain). Note: In other hash-chain based systems Holochain's "actions" are often called entry "headers," which link to the previous headers to create the chain.  In Holochain, while the action contains the provenance and establishes temporal order, their core function is to record an act of agency, that of "speaking" data into existence.
```

Comment:
> entries must carry the associated action

This isn't true since Entries in general don't have a single Action. Also, it's very possible by design to access an Entry without regard to any Actions that created it. Entries don't "carry" anything but themselves. This paragraph should be rewritten to account for that.